### DISCUSSION TITLE
Why box the handler?

### DISCUSSION THREAD

**Discussion by kingluo** [routerify/routerify#83] [Q&A]
https://github.com/routerify/routerify/blob/master/src/route/mod.rs#L89

**Comment by gsserge** [comment]
Because `async fn` is transformed by the compiler to a function that returns `impl Future`.  The type of `impl Future` is anonymous (known only to the compiler) and unique - it's a generated state machine enum behind the scenes. So when you want to use a container to store multiple handlers, a hash map or a vector, the value type must be the same. And that's not the case because
```rust
async fn hello(req: http::Request) -> http::Response {}
```
and 
```rust
async fn bye(req: http::Request) -> http::Response {}
```
are of different types despite the fact that the signatures look the same. You are forced to use trait objects instead, and that's why you need to box the handlers.

**Reply by kingluo** [reply]
It makes sense, but why two `Box`?
`Box::new(move |req: Request<hyper::Body>| Box::new(handler(req)))`
